main/verifier.go

```diff
@@ -135,54 +62,113 @@ func loadUPP(hashString string, conf Config) ([]byte, int, error) {
 			}
 		}
 	}
+	//noinspection GoUnhandledErrorResult
+	defer resp.Body.Close()
 
 	if resp.StatusCode != http.StatusOK {
-		return nil, resp.StatusCode, fmt.Errorf("could not (yet) retrieve certificate for hash %s from verification service (%s): %s", hashString, verificationURL, resp.Status)
+		return nil, resp.StatusCode, fmt.Errorf("could not (yet) retrieve certificate for hash %s from verification service (%s): %s", hashString, verifyService, resp.Status)
 	}
 
 	vf := verification{}
 	err = json.NewDecoder(resp.Body).Decode(&vf)
 	if err != nil {
 		return nil, http.StatusInternalServerError, fmt.Errorf("unable to decode verification response: %v", err)
 	}
-	_ = resp.Body.Close()
-	return vf.UPP, resp.StatusCode, nil
+	return vf.UPP, http.StatusOK, nil
+}
+
+// verifyUPP verifies the validity of a UPP's signature. Requests public key from the key service if unknown.
+func verifyUPP(p *ExtendedProtocol, upp []byte, keyService string) (uuid.UUID, error) {
+	o, err := ubirch.DecodeChained(upp)
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("UPP decoding failed: %v", err)
+	}
+	id := o.Uuid
+	name := id.String()
+
+	pubkey, err := loadPublicKey(p, name, id, keyService)
+	if err != nil {
+		return id, err
+	}
+	log.Debugf("verifying validity of UPP signature using pubkey %s of identity %s", base64.StdEncoding.EncodeToString(pubkey), name)
+
+	verified, err := p.Verify(name, upp, ubirch.Chained)
+	if err != nil {
+		return id, err
+	}
+	if !verified {
+		return id, fmt.Errorf("validity of UPP signature could not be verified using public key %s of identity %s", base64.StdEncoding.EncodeToString(pubkey), name)
+	}
+	return id, nil
 }
 
-// hash a message and retrieve corresponding UPP to verify it
+// loadPublicKey retrieves the first valid public key associated with an identity from the key service
```

Comment:
what happens, if the first public key is not the right one?

Reply:
That is a known issue and needs some thought/work. currently there will only be a warning saying "more than one key registered" and the client will return an error if validation with first public key fails.